types/validation.go

```diff
@@ -273,11 +279,19 @@ func verifyCommitBatch(
 		// Validate signature.
 		voteSignBytes := commit.VoteSignBytes(chainID, int32(idx))
 
-		// add the key, sig and message to the verifier
-		if err := bv.Add(val.PubKey, voteSignBytes, commitSig.Signature); err != nil {
-			return err
+		cacheKey, cacheValAddress, sigIsInCache := "", []byte{}, false
+		if verifiedSignatureCache != nil {
+			cacheKey = string(voteSignBytes)
+			cacheValAddress, sigIsInCache = verifiedSignatureCache[cacheKey]
+		}
+
+		if !sigIsInCache || !bytes.Equal(cacheValAddress, commitSig.ValidatorAddress) {
```

Comment:
Can we make a single boolean for "cacheHit"

Reply:
Good call, done.

Will work on adding tests.